components/_impl/workers/subprocess_rpc.py

```diff
@@ -102,17 +104,19 @@ class _TimeoutPIPE:
     _loop_lock = threading.Lock()
     _active_reads: typing.Dict[int, typing.Tuple[float, float]]
     _loop_cadence = 1  # second
+    _proc_id: typing.Optional[int] = None
 
     @classmethod
-    def singleton(cls) -> "_TimeoutPIPE":
+    def singleton(cls, proc_id) -> "_TimeoutPIPE":
         # This class will spawn a thread, so we only want one active at a time.
         with cls._singleton_lock:
             if cls._singleton is None:
-                cls._singleton = cls()
+                cls._singleton = cls(proc_id)
             return cls._singleton
 
-    def __init__(self) -> None:
+    def __init__(self, proc_id: int) -> None:
         self._active_reads = {}
```

Comment:
If you store the pid as a third element of `_active_reads`, you can associate a pid per outstanding read and not have to specialize the class itself to a particular child.